src/mds/CInode.h

```diff
@@ -1089,28 +1092,51 @@ class CInode : public MDSCacheObject, public InodeStoreBase, public Counter<CIno
   elist<CInode*>::item item_pop_lru;
 
   // -- locks --
-  static LockType versionlock_type;
-  static LockType authlock_type;
-  static LockType linklock_type;
-  static LockType dirfragtreelock_type;
-  static LockType filelock_type;
-  static LockType xattrlock_type;
-  static LockType snaplock_type;
-  static LockType nestlock_type;
-  static LockType flocklock_type;
-  static LockType policylock_type;
-
-  // FIXME not part of mempool
-  LocalLockC  versionlock;
-  SimpleLock authlock;
-  SimpleLock linklock;
-  ScatterLock dirfragtreelock;
-  ScatterLock filelock;
-  SimpleLock xattrlock;
-  SimpleLock snaplock;
-  ScatterLock nestlock;
-  SimpleLock flocklock;
-  SimpleLock policylock;
+  static const LockType versionlock_type;
+  static const LockType authlock_type;
+  static const LockType linklock_type;
+  static const LockType dirfragtreelock_type;
+  static const LockType filelock_type;
+  static const LockType xattrlock_type;
+  static const LockType snaplock_type;
+  static const LockType nestlock_type;
+  static const LockType flocklock_type;
+  static const LockType policylock_type;
+  static const LockType quiescelock_type;
+
+  /* A note about quiescelock:
+   *
+   * It is a type of superlock where every operation which accesses an Inode
+   * lock will also acquire the quiescelock (readonly). In general, this lock
+   * is never held except for reading. When a subtree is quiesced, the
+   * quiesce_subvolume_inode internal operation will hold quiescelock
+   * exclusively, thereby denying any other locks. The quiescelock must be
+   * ordered before all other locks (see src/include/ceph_fs.h for ordering) in
+   * order to act as this superlock.
+   *
+   * It may be desirable to allow readers to continue accessing a quiesced
+   * subvolume. One way to do that is to have a separate superlock for read
+   * access, say "quiescerlock". If a "readable" quiesce is performed, then
+   * "quiescerlock" is not acquired by quiesce_subvolume_inode. Read locks on
+   * other (non-quiesce) locks will acquire a read lock only on "quiescerlock"
+   * and not on "quiescelock". Write locks would try to acquire both
+   * quiescelock and quiescerlock (since writes may also read).
+   *
+   * Ideally, it may be a new lock could be used to handle both cases but no
```

Comment:
what about the FileLock type and its three levels of `rdlock,wrlock,xlock`. Couldn't we utilize this here?

Reply:
looking further, I really like the idea of adding another lock class for the quiesce lock. It could be a rather simple class allowing the readable quiesce semantics

Reply:
The `FileLock` is unfortunately not appropriate. It allows a reader/writer mix. It is more like, `rwlock, xlock`. i.e. acquiring a `rdlock` on the `filelock` does not prevent writers. What we need is a lock which is `rdonly` ("readable" quiesce), `multi rd/wr` (normal operation), or `xlock` (exclusive quiesce).